src/TheFox/Smtp/Server.php

```diff
@@ -252,4 +262,28 @@ public function mailNew($from, $rcpt, $mail){
 		$this->eventExecute(Event::TRIGGER_MAIL_NEW, array($from, $rcpt, $mail));
 	}
 	
+	/**
+	 * Execute authentication events
+	 * 
+	 * All authentication events must return true for authentication to be successful
+	 * 
+	 */
+	public function authenticateUser($method, $credentials = array()){
+		$authenticated = false;
+		$args = array($method, $credentials);
+		
+		foreach($this->events as $eventId => $event){
+			if($event->getTrigger() == Event::TRIGGER_AUTH_ATTEMPT){
+				if($event->execute($args)){
+					$authenticated = true;
+				}
+				else{
+					$authenticated = false;
```

Comment:
But your code take only the last event. If the last event returns `false` the function also returns `false`, no matter the other events. Think about this code instead:

```php
public function authenticateUser($method, $credentials = array()){
	$args = array($method, $credentials);
	
	foreach($this->events as $eventId => $event){
		if($event->getTrigger() == Event::TRIGGER_AUTH_ATTEMPT){
			if(!$event->execute($args)){
				return false;
			}
		}
	}
	
	return true;
}
```

It returns `true` by default but in the case one of the events fails the function also returns `false`. In other words, if every event is `true` the function is also `true`.

But this is actually also **wrong**. Consider you don't have any `Event::TRIGGER_AUTH_ATTEMPT` events, this function always returns `true`. So we also need to check if we have at least one `Event::TRIGGER_AUTH_ATTEMPT` event. So I would rather use this code instead:

```php
public function authenticateUser($method, $credentials = array()){
	$hasAtLeastOneAuthAttempEvent = false;
	$args = array($method, $credentials);
	
	foreach($this->events as $eventId => $event){
		if($event->getTrigger() == Event::TRIGGER_AUTH_ATTEMPT){
			if(!$event->execute($args)){
				return false;
			}
			$hasAtLeastOneAuthAttempEvent = true;
		}
	}
	
	return $hasAtLeastOneAuthAttempEvent;
}
```

This function returns `true` only if at least one `Event::TRIGGER_AUTH_ATTEMPT` was in the events array AND if all `Event::TRIGGER_AUTH_ATTEMPT` events were `true`. Otherwise we return `false` on the first event that returned `false`